chore: add local development with SDK backends guide (#261)
kubeflow/sdk: docs/source/getting-started/local-development.rst

Local Development
==================

This guide explains how to run Kubeflow TrainJobs locally using various Kubeflow SDK backends, allowing for faster iteration before deploying to a Kubernetes cluster.

> Overview

Reviewer: RST needs a blank line between the intro paragraph above and this heading, otherwise `Overview` won't render as a section header.

```suggestion

Overview
```

--------

The Kubeflow SDK provides three backends for running TrainJobs:

.. list-table:: Backend Comparison
   :header-rows: 1
   :widths: 20 35 45

   * - Backend
     - Best For
     - Requirements
   * - **Local Process**
     - Quick prototyping, single-node testing

>      - Python 3.9+

Reviewer: `pyproject.toml` requires `python >= 3.10`. Please update to `Python 3.10+`.

   * - **Container**

>      - Multi-node training, reproducibility

Reviewer: The Container backend launches multiple containers on a single host, not across physical nodes. Consider rewording to `Multi-process distributed training (single host), reproducibility` to avoid implying multi-machine support.

     - Docker or Podman installed
   * - **Kubernetes**
     - Production deployments
     - K8s cluster with Trainer operator

All backends use the same ``TrainerClient`` interface - only the configuration
changes. This means you can develop locally and deploy to production with
minimal code changes.

Local Process Backend
---------------------

The fastest option for quick testing. Runs training directly as Python processes.

**When to use:**

- Rapid prototyping and debugging
- Testing training logic without container overhead
- Environments without Docker/Podman

**Example:**

.. code-block:: python

   from kubeflow.trainer import TrainerClient, LocalProcessBackendConfig
   from kubeflow.trainer import CustomTrainer

   # Configure local process backend
   backend_config = LocalProcessBackendConfig()
   client = TrainerClient(backend_config=backend_config)

   # Define your training function
   def train_model():
       import torch

>        print(f"Training on device: {torch.cuda.current_device() if torch.cuda.is_available() else 'cpu'}")

Reviewer: `torch.cuda.current_device()` returns the device index (e.g. `0`), which is inconsistent with the `'cpu'` string. Suggest:

```suggestion
       print(f"Training on device: {'cuda' if torch.cuda.is_available() else 'cpu'}")
```

       # Your training logic here

   # Create trainer and run
   trainer = CustomTrainer(func=train_model)
   job_name = client.train(trainer=trainer)

   # View logs
   client.get_job_logs(name=job_name, follow=True)

**Limitations:**

- Single-node only (no distributed training)
- No container isolation

Container Backend (Docker/Podman)
---------------------------------

> Run training in isolated containers with full multi-node distributed training support.

Reviewer: Same concern as the comparison table: the Container backend runs multiple containers on one host, not multi-node. Suggest rewording to avoid implying multi-machine distributed training.

**When to use:**

- Distributed training with multiple workers
- Reproducible containerized environments
- Testing production-like setups locally

**Example with Docker:**

.. code-block:: python

   from kubeflow.trainer import TrainerClient, ContainerBackendConfig
   from kubeflow.trainer import CustomTrainer

   # Configure Docker backend
   backend_config = ContainerBackendConfig(
       container_runtime="docker",  # or "podman"
   )
   client = TrainerClient(backend_config=backend_config)

   # Same trainer works - now with multi-node support!
   trainer = CustomTrainer(
       func=train_model,
       num_nodes=4,  # Distributed across 4 containers
   )
   job_name = client.train(trainer=trainer)

.. _container-host-configuration:

Container Host Configuration
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> When using the Container backend on **macOS** , you may need to configure the

Reviewer: Stray space before the comma.

```suggestion
When using the Container backend on **macOS**, you may need to configure the
```

``container_host`` parameter to point to your Docker or Podman socket. This is
because the default socket path differs across operating systems.

.. list-table::
   :header-rows: 1
   :widths: 20 80

   * - OS
     - Default ``container_host``
   * - Linux

>      - ``unix:///var/run/docker.sock`` (Docker) or ``unix:///run/user/<UID>/podman/podman.sock`` (Podman)

Reviewer: Worth clarifying for readers that `<UID>` is the current user's ID (from `id -u`), otherwise it looks like a literal value to substitute.

   * - macOS
     - ``unix://$HOME/.docker/run/docker.sock`` (Docker Desktop) or check ``podman machine inspect`` for Podman
   * - Windows
     - ``npipe:////./pipe/docker_engine`` (Docker Desktop)

**Example for macOS:**

.. code-block:: python

   import os

   backend_config = ContainerBackendConfig(
       container_runtime="docker",
       # macOS Docker Desktop socket path
       container_host=f"unix://{os.environ['HOME']}/.docker/run/docker.sock",
   )
   client = TrainerClient(backend_config=backend_config)

.. note::

   If you encounter ``Cannot connect to Docker daemon`` errors on macOS,
   verify the socket path by running ``docker context inspect`` and check
   the ``Host`` value in the output.

> 

Reviewer: Consider adding a dedicated `Kubernetes Backend` section here, parallel to `Local Process Backend` and `Container Backend`. Kubernetes is listed in the comparison table but never gets its own section, leaving the doc structurally incomplete.

Switching Between Backends
--------------------------

A key benefit of the Kubeflow SDK is seamless backend switching. Your training code
stays the same - only the backend configuration changes:

.. code-block:: python

   # Development: Use local process for fast iteration
   from kubeflow.trainer import LocalProcessBackendConfig
   backend_config = LocalProcessBackendConfig()

   # Testing: Switch to Docker for distributed testing
   from kubeflow.trainer import ContainerBackendConfig
   backend_config = ContainerBackendConfig(container_runtime="docker")

   # Production: Deploy to Kubernetes
   from kubeflow.trainer import KubernetesBackendConfig

>    backend_config = KubernetesBackendConfig(namespace="kubeflow")

Reviewer: This is the only place `KubernetesBackendConfig` appears — consider mentioning other commonly-used fields (`config_file`, `context`) so users know the full surface area.

Local Development
==================

This guide explains how to run Kubeflow TrainJobs locally using the SDK's
different backends, helping you iterate faster before deploying to a Kubernetes
cluster.

Overview
--------

> The Kubeflow Trainer SDK provides three backends for running TrainJobs:

Reviewer 2: ```suggestion
The Kubeflow SDK provides three backends for running TrainJobs:
```

because the default socket path differs across operating systems.

.. list-table::
   :header-rows: 1
   :widths: 20 80

   * - OS
     - Default ``container_host``
   * - Linux
     - ``unix:///var/run/docker.sock`` (Docker) or ``unix:///run/user/<UID>/podman/podman.sock`` (Podman)
   * - macOS
     - ``unix://$HOME/.docker/run/docker.sock`` (Docker Desktop) or check ``podman machine inspect`` for Podman
   * - Windows
     - ``npipe:////./pipe/docker_engine`` (Docker Desktop)

**Example for macOS:**

.. code-block:: python

   import os

   backend_config = ContainerBackendConfig(
       container_runtime="docker",
       # macOS Docker Desktop socket path
       container_host=f"unix://{os.environ['HOME']}/.docker/run/docker.sock",
   )
   client = TrainerClient(backend_config=backend_config)

.. note::

   If you encounter ``Cannot connect to Docker daemon`` errors on macOS,
   verify the socket path by running ``docker context inspect`` and check
   the ``Host`` value in the output.

Switching Between Backends
--------------------------

The key benefit of the SDK is seamless backend switching. Your training code

> stays the same - only the backend configuration changes:

Reviewer 2: ```suggestion
A key benefit of the Kubeflow SDK is seamless backend switching. Your training code
stays the same - only the backend configuration changes:
```

Local Development with SDK Backends
====================================

This guide explains how to run Kubeflow training jobs locally using the SDK's
different backends, helping you iterate faster before deploying to a Kubernetes
cluster.

Overview
--------

> The Kubeflow Trainer SDK provides three backends for running training jobs:

Reviewer 2: ```suggestion
The Kubeflow Trainer SDK provides three backends for running TrainJobs:
```
